Feature/termux controls provider

dkramer95 opened the pull request:
Implements #54 Device Control Widgets
**(This is only for Android 11+)**
I have only tested on Google Pixel 5

![Screen Shot 2021-08-29 at 1 38 43 PM](https://user-images.githubusercontent.com/6166095/131261657-e686d41c-afb2-453d-97c8-ebc781d794e3.png)

Execution will exhibit same behavior as current home screen widgets. (i.e. any file placed in `.shortcuts/tasks` will run in background, all others in foreground.
* Exception: If script runs any ui action (ex: `termux-dialog`, `termux-toast`) it will not be visible until user exits the Device Control view.

## Error handling if script was removed / changed:

* Widget will show "Not found". Tapping on it will display this error dialog:

![Screen Shot 2021-08-29 at 1 40 18 PM](https://user-images.githubusercontent.com/6166095/131261710-f4a52b12-5095-4842-8ef5-5ee3876a0671.png)
and tapping "Open App" will display error in Termux window:
![Screen Shot 2021-08-29 at 1 41 21 PM](https://user-images.githubusercontent.com/6166095/131261734-89bae65b-faef-4e6c-b2ae-a2320e72146e.png)

agnostic-apollo reviewed (changes requested):
>`private Intent addShortcutFileExtrasToIntent(File file, Intent intent)`

Its returning an `Intent` but its never used. You are passing the intent as parameter which will automatically get updated, so no need to return.

> `private void addShortcutFile(File dir, List<File> shortcutFiles, int depth)`

Why are you not using `dir.listFiles(TermuxWidgetService.SHORTCUT_FILES_FILTER)`? Wouldn't shortcut icons and hidden files be included currently?

Apart from that, good coding! I don't *see* any other issues related to new changes. I can't test this cause my Android `11` `avd` is not showing the 3 dots for me to add the device control in power menu, it seems android revision is old, probably that's the reason.

Also note that in Android `12` device controls have been removed from power menu. Check https://www.reddit.com/r/Android/comments/nx9pd1/android_12_device_controls_removed_from_the_power. No response from google in https://issuetracker.google.com/issues/190662102.

dkramer95 commented:
> Its returning an Intent but its never used. You are passing the intent as parameter which will automatically get updated, so no need to return.

Updated!

> Why are you not using dir.listFiles(TermuxWidgetService.SHORTCUT_FILES_FILTER)? Wouldn't shortcut icons and hidden files be included currently?

Fixed.. Somehow I didn't see that was there :)

> Also note that in Android 12 device controls have been removed from power menu.

This would be a very niche feature then. Hopefully at the very least, Google will allow an option to re-enable it. 🤞

Berat7171 commented:
55
#55

Berat7171 commented:
İnstall
